packages/components/src/navigation/types.ts

```diff
@@ -0,0 +1,316 @@
+	 * `navigateToMenu` is marked as empty in the `isEmpty` prop.
+	 * Used together with the `isEmpty` prop of `NavigationMenu`.
+	 */
+	hideIfTargetMenuEmpty?: boolean;
+	/**
+	 * A callback to handle clicking on a menu item.
+	 *
+	 * @default noop
+	 */
+	onClick?: React.MouseEventHandler;
+	/**
+	 * The item title.
+	 */
+	title?: string;
+};
+
+export type NavigationItemProps = _NavigationItemButtonProps &
+	NavigationItemBaseProps;
+
+export type NavigationItemBaseContentProps = Pick<
+	NavigationItemProps,
+	'badge' | 'title'
+>;
+
+export type NavigationBackButtonProps = {
+	/**
+	 * The text displayed for the button
+	 */
+	backButtonLabel?: string;
+	/**
+	 * Optional classname for the component.
+	 */
+	className?: string;
+	/**
+	 * If provided, causes the component to render an `<a />` element
+	 * instead of a `<button />` element.
+	 */
+	href?: string;
+	/**
+	 * Callback function called when the button is clicked.
+	 */
+	onClick?: React.MouseEventHandler;
+	/**
+	 * Id of the parent menu
+	 */
+	parentMenu?: string;
+};
+
+export type NavigationMenuTitleSearchProps = Pick<
+	NavigationMenuProps,
+	'onSearch' | 'search' | 'title'
+> & {
+	// TODO: grab types from useDebounce / speak ?
+	debouncedSpeak: ( message: string ) => void;
+	onCloseSearch: () => void;
+};
+
+export type NavigationMenuTitleProps = Pick<
+	NavigationMenuProps,
+	'hasSearch' | 'onSearch' | 'search' | 'title' | 'titleAction'
+>;
+
+export type NavigationSearchNoResultsFoundProps = Pick<
+	NavigationMenuProps,
+	'search'
+>;
+
+// Data objects & context
+
+export type Item = NavigationItemBaseProps & {
+	/**
+	 * The group id
+	 */
+	group?: string;
+	/**
+	 * The menu id
+	 */
+	menu?: string;
+	/**
+	 * True is the item is currently visible (ie. the item is part
+	 * of the active menu, and — if there is a search query — its title
+	 * matches the search query).
+	 */
+	_isVisible: boolean;
+};
+
+export type Menu = NavigationMenuProps & {
+	menu: string;
+};
+
+export type NavigationGroupContext = {
+	/**
+	 * The unique id of the group
+	 */
+	group?: string;
+};
+
+export type NavigationMenuContext = {
+	menu?: string;
+	search?: string;
+};
+
+export type NavigationContext = {
+	activeItem?: string;
+	activeMenu: string;
+	setActiveMenu: (
+		parentMenu: string,
+		animationDirection?: 'left' | 'right'
+	) => void;
+
+	navigationTree: {
+		items: Record< string, Item >;
+		getItem: ( itemId: string ) => Item | undefined;
+		addItem: ( itemId: string, itemProps: Item ) => void;
+		removeItem: ( itemId: string ) => void;
+
+		menus: Record< string, Menu >;
+		getMenu: ( menuId: string ) => Menu | undefined;
+		addMenu: ( menuId: string, menuProps: Menu ) => void;
+		removeMenu: ( menuId: string ) => void;
+		childMenu: Record< string, string[] >;
+		traverseMenu: (
+			startMenu: string,
+			callback: ( menuObject: Menu ) => boolean | undefined
+		) => void;
+		isMenuEmpty: ( menuId: string ) => boolean;
```

Comment:
Any concerns that those types don't 100% match the default context? For example, `defaultIsEmpty` expects no args, but the corresponding `isMenuEmpty` type expects a `menuId`.

Reply:
It shouldn't be an issue — the source of truth should be the types.

The fact that the `defaultIsEmpty` function doesn't accept any argument shouldn't change the behaviour of the function, which will still return a `boolean` nevertheless

```js
defaultIsEmpty(); // `false`
defaultIsEmpty( 'a string' ); // `false`
defaultIsEmpty( undefined ) // `false`
defaultIsEmpty( null ) // `false`
```

This non-breaking behaviour should be further certified by the lack of any TypeScript errors.


This is the same reason why it's usually not an issue that we sometimes assign `noop` as a default value of a callback, regardless of the callback signature. (even though I'm not a fan of that pattern myself).

Reply:
Sounds good to me, thanks for confirming 👍